File: src/lib/gemini.ts
import {
  GoogleGenerativeAI,
  TaskType,
  type FunctionDeclarationsTool,
  type GenerateContentResult,
} from '@google/generative-ai'

const genAI = new GoogleGenerativeAI(process.env.GOOGLE_API_KEY || '')

export const embeddingModel = genAI.getGenerativeModel({
  model: 'gemini-embedding-001',
})

export const chatModel = genAI.getGenerativeModel({
  model: 'gemini-2.0-flash',
})

// Retry wrapper for Gemini 429 rate limits
async function withRetry<T>(fn: () => Promise<T>, maxRetries = 3): Promise<T> {
  for (let attempt = 0; attempt < maxRetries; attempt++) {
    try {
      return await fn()
    } catch (error: any) {
      const status = error?.status || error?.httpStatusCode
      if (status === 429 && attempt < maxRetries - 1) {
        const delay = (attempt + 1) * 5000 // 5s, 10s, 15s
        console.log(`[GEMINI] Rate limited, retrying in ${delay / 1000}s (attempt ${attempt + 1}/${maxRetries})`)
        await new Promise((r) => setTimeout(r, delay))
        continue
      }
      throw error
    }
  }
  throw new Error('Max retries exceeded')
}

export async function generateEmbedding(
  text: string,
  taskType: 'retrieval_document' | 'retrieval_query' = 'retrieval_document'
): Promise<number[]> {
  const tt =
    taskType === 'retrieval_document'
      ? TaskType.RETRIEVAL_DOCUMENT
      : TaskType.RETRIEVAL_QUERY

  return withRetry(async () => {
    const result = await embeddingModel.embedContent({
      content: { parts: [{ text }], role: 'user' },
      taskType: tt,
    })
    return result.embedding.values
  })
}

export async function generateChatResponse(
  systemPrompt: string,
  userMessage: string,
  conversationHistory: { role: 'user' | 'model'; content: string }[] = []
): Promise<string> {
  return withRetry(async () => {
    const chat = chatModel.startChat({
      systemInstruction: { role: 'user', parts: [{ text: systemPrompt }] },
      history: conversationHistory.map((msg) => ({
        role: msg.role,
        parts: [{ text: msg.content }],
      })),
    })

    const result = await chat.sendMessage(userMessage)
    return result.response.text()
  })
}

export async function checkQueryAccess(
  query: string,
  userTitle: string
): Promise<{ allowed: boolean; reason?: string }> {
  return withRetry(async () => {
    const designationLabel = userTitle || 'employee'
    const prompt = `You are an access control checker. A user is asking a question. Determine if the query is asking about a specific designation band that DOES NOT match their profile.

User profile:
- Designation/Title: ${designationLabel}

Designation bands:
- Band A (Senior Leadership): Director, Senior Director, Vice President, CXO
- Band B (Mid-Level Management): Manager, Senior Manager, Lead Architect, Program Manager
- Band C (Individual Contributors): Engineer, Analyst, Associate, Executive

User query: "${query}"

Rules:
- If the query explicitly asks about a DIFFERENT designation band (e.g. user is an Analyst but asks about "Director benefits" or "VP reimbursement limits"), respond: DENIED|You don't have access to that designation band's information. Your designation is ${designationLabel}.
- If the query is general (e.g. "what is the mobile reimbursement policy?") or matches the user's designation band, respond: ALLOWED
- If unclear, respond: ALLOWED

Respond with ONLY "ALLOWED" or "DENIED|reason". Nothing else.`

    const result = await chatModel.generateContent(prompt)
    const response = result.response.text().trim()

    if (response.startsWith('DENIED')) {
      const reason = response.split('|')[1] || "You don't have access to that information."
      return { allowed: false, reason }
    }
    return { allowed: true }
  })
}

export async function summarizeArticle(
  title: string,
  body: string
): Promise<string> {
  return withRetry(async () => {
    const prompt = `You are an enterprise knowledge base article processor. Given a raw article (often with HTML markup), produce a clean, well-structured summary that preserves ALL important information.

Think step by step:
1. First, strip all HTML tags and formatting noise
2. Identify every factual detail, number, policy rule, eligibility criteria, deadline, limit, exception, and condition
3. Organize the information logically with clear sections and hierarchy
4. Preserve all specific values (amounts, dates, percentages, band/tier details, designation-specific rules)
5. Keep role/designation-specific distinctions clear (e.g., different limits for different bands)

Title: ${title}

Raw Article:
${body.substring(0, 10000)}

Rules:
- Do NOT lose any factual information — every number, rule, condition, eligibility, and exception must be retained
- Use clean markdown formatting (headers, bullet points, sub-bullets)
- Remove HTML tags, redundant whitespace, and formatting artifacts
- Keep the language professional and concise but complete
- Structure with clear sections matching the original article structure
- If the article has designation/role/band-specific info, keep each band's details clearly separated

Respond with ONLY the processed article content. No preamble.`

    const result = await chatModel.generateContent(prompt)
    return result.response.text().trim()
  })
}

export async function generateWithTools(
  systemPrompt: string,
  userMessage: string,
  tools: FunctionDeclarationsTool[],
  modelName = 'gemini-2.0-flash',
): Promise<GenerateContentResult> {
  return withRetry(async () => {
    const model = genAI.getGenerativeModel({
      model: modelName,
      systemInstruction: { role: 'user', parts: [{ text: systemPrompt }] },
      tools,
    })
    return model.generateContent(userMessage)
  })
}

export async function generateWithToolsChat(
  systemPrompt: string,
  messages: { role: 'user' | 'model' | 'function'; parts: any[] }[],
  tools: FunctionDeclarationsTool[],
  modelName = 'gemini-2.0-flash',
): Promise<GenerateContentResult> {
  return withRetry(async () => {
    const model = genAI.getGenerativeModel({
      model: modelName,
      systemInstruction: { role: 'user', parts: [{ text: systemPrompt }] },
      tools,
    })

    const chat = model.startChat({
      history: messages.slice(0, -1),
    })

    const lastMessage = messages[messages.length - 1]
    return chat.sendMessage(lastMessage.parts)
  })
}

export async function classifyIntent(
  userMessage: string
): Promise<'kb_query' | 'conversational' | 'workflow'> {
  return withRetry(async () => {
    const prompt = `You are an intent classifier. Classify the following user message into exactly one category:

1. "kb_query" - The user is asking a question that can be answered from a knowledge base (policies, procedures, FAQs, how-to guides, holiday calendars, leave policies, hardware procurement, performance reviews, etc.)
2. "conversational" - The user is making small talk, greetings, or asking something that doesn't require knowledge base lookup
3. "workflow" - The user wants to trigger an action/workflow (create ticket, request approval, etc.)

User message: "${userMessage}"

Respond with ONLY the category name, nothing else.`

    const result = await chatModel.generateContent(prompt)
    const response = result.response.text().trim().toLowerCase()

    if (response.includes('kb_query')) return 'kb_query'
    if (response.includes('workflow')) return 'workflow'
    return 'conversational'
  })
}

File: backend/services/gemini.py
import asyncio
import json
import logging
from google import genai
from google.genai import types
from config import get_settings
from services.slm import slm_generate, strip_thinking

logger = logging.getLogger(__name__)

settings = get_settings()
client = genai.Client(api_key=settings.google_api_key)

EMBEDDING_MODEL = "gemini-embedding-001"
# gemini-2.5-flash-lite has a higher free-tier daily quota than gemini-2.5-flash,
# which makes the multi-call agentic workflows usable without a paid key.
CHAT_MODEL = "gemini-2.5-flash-lite"


async def with_retry(fn, max_retries: int = 3):
    """Retry wrapper for Gemini 429 rate limits."""
    for attempt in range(max_retries):
        try:
            return await fn()
        except Exception as e:
            status = getattr(e, "status", None) or getattr(e, "code", None)
            if status == 429 and attempt < max_retries - 1:
                delay = (attempt + 1) * 5
                await asyncio.sleep(delay)
                continue
            raise
    raise Exception("Max retries exceeded")


async def generate_embedding(
    text: str,
    task_type: str = "RETRIEVAL_DOCUMENT",
) -> list[float]:
    """Generate a 3072-dim embedding for text."""
    async def _call():
        result = await asyncio.to_thread(
            client.models.embed_content,
            model=EMBEDDING_MODEL,
            contents=text,
            config=types.EmbedContentConfig(task_type=task_type),
        )
        return result.embeddings[0].values

    return await with_retry(_call)


async def generate_chat_response(
    system_prompt: str,
    user_message: str,
    conversation_history: list[dict] | None = None,
) -> str:
    """Generate a chat response with optional conversation history."""
    async def _call():
        contents = []
        if conversation_history:
            for msg in conversation_history:
                role = "user" if msg["role"] == "user" else "model"
                contents.append(types.Content(role=role, parts=[types.Part(text=msg["content"])]))
        contents.append(types.Content(role="user", parts=[types.Part(text=user_message)]))

        response = await asyncio.to_thread(
            client.models.generate_content,
            model=CHAT_MODEL,
            contents=contents,
            config=types.GenerateContentConfig(
                system_instruction=system_prompt,
            ),
        )
        return response.text

    return await with_retry(_call)


async def generate_with_tools(
    system_prompt: str,
    user_message: str,
    tools: list[types.Tool],
    model_name: str = CHAT_MODEL,
) -> types.GenerateContentResponse:
    """Generate content with function calling tools."""
    async def _call():
        return await asyncio.to_thread(
            client.models.generate_content,
            model=model_name,
            contents=user_message,
            config=types.GenerateContentConfig(
                system_instruction=system_prompt,
                tools=tools,
            ),
        )

    return await with_retry(_call)


async def generate_with_tools_chat(
    system_prompt: str,
    messages: list[types.Content],
    tools: list[types.Tool],
    model_name: str = CHAT_MODEL,
) -> types.GenerateContentResponse:
    """Generate content with tools using chat history."""
    async def _call():
        return await asyncio.to_thread(
            client.models.generate_content,
            model=model_name,
            contents=messages,
            config=types.GenerateContentConfig(
                system_instruction=system_prompt,
                tools=tools,
            ),
        )

    return await with_retry(_call)


async def classify_intent(user_message: str) -> str:
    """Classify user intent: kb_query, conversational, or workflow."""
    prompt = f"""You are an intent classifier for an enterprise AI assistant. Classify the following user message into exactly one category:

1. "kb_query" - The user is asking a question that can be answered from a knowledge base (policies, procedures, FAQs, how-to guides, holiday calendars, leave policies, hardware procurement, performance reviews, company info, etc.)
2. "conversational" - The user is making small talk, greetings, thank you, or asking something generic that doesn't require knowledge base lookup or any action
3. "workflow" - The user wants to perform an action or trigger a workflow. Examples:
   - Create/update/close a ServiceNow incident or ticket
   - Create/update a Jira issue or ticket
   - Send a Slack message or notification
   - Manage AWS resources (EC2 instances, S3 buckets)
   - Upload/search files in SharePoint
   - Run a database query in Snowflake
   - Any request that involves doing something, creating something, sending something, or modifying external systems

User message: "{user_message}"

Respond with ONLY the category name, nothing else."""

    # Try SLM first (fast, local classification)
    try:
        slm_prompt = f"/no_think\n{prompt}"
        response = await slm_generate(
            "You are an intent classifier. Respond with only one word.",
            slm_prompt,
            temperature=0.1,
            max_tokens=20,
        )
        text = strip_thinking(response).strip().lower()
        if "kb_query" in text:
            return "kb_query"
        if "workflow" in text:
            return "workflow"
        return "conversational"
    except Exception as e:
        logger.warning("SLM classify_intent failed, falling back to Gemini: %s", e)

    # Gemini fallback
    async def _call():
        response = await asyncio.to_thread(
            client.models.generate_content,
            model=CHAT_MODEL,
            contents=prompt,
        )
        text = response.text.strip().lower()
        if "kb_query" in text:
            return "kb_query"
        if "workflow" in text:
            return "workflow"
        return "conversational"

    return await with_retry(_call)


async def summarize_article(title: str, body: str) -> str:
    """Summarize a knowledge base article."""
    prompt = f"""You are an enterprise knowledge base article processor. Given a raw article (often with HTML markup), produce a clean, well-structured summary that preserves ALL important information.

Title: {title}

Raw Article:
{body[:10000]}

Rules:
- Do NOT lose any factual information
- Use clean markdown formatting
- Remove HTML tags
- Keep designation/role-specific info clearly separated

Respond with ONLY the processed article content. No preamble."""

    async def _call():
        response = await asyncio.to_thread(
            client.models.generate_content,
            model=CHAT_MODEL,
            contents=prompt,
        )
        return response.text.strip()

    return await with_retry(_call)
